Bluetooth: Fix adv scan random addr

joerchan opened the pull request:
Handle multirole scenarios where multiple roles might use the same random address
in the controller, but using different identities.

Handle scanner active at RPA timeout:
Fix RPA timeout handling when the scanner is active. An active scanner
must be restarted at RPA timeout otherwise the Set Random Address
command will fail.

Handle starting roles with different random address:
Handle starting of advertiser and scanner or initiator when advertiser
is using a different identity than the default identity to generate the
random resolvable address in the controller.
We need to handle this only for the privacy case because the random
address is set in the RPA timeout handler and not from the API.
When privacy is disabled we can return error code from the LE Set Random
Address HCI command instead.

Handle initiator at RPA timeout:
Handle initiator role when RPA timeout expires. For direct connect
establishment procedure we make sure the RPA is refreshed when starting
initiator and limit the timeout to the RPA timeout.
For auto establishment procedure we cancel the initiator and restart it
again in the connection complete event that is generated when canceling
an initiator.

Handle advertiser with Identity switches to RPA
Fix advertiser requested to use the identity address while privacy
feature is enabled will change to using RPA address when advertise is
resumed or when RPA timeout occurred.
RPA timeout does not need to run when advertiser is using identity.

joerchan committed:
Bluetooth: host: Handle starting roles with different random address

Handle starting of advertiser and scanner or initiator when advertiser
is using a different identity than the default identity to generate the
random resolvable address in the controller.
We need to handle this only for the privacy case because the random
address is set in the RPA timeout handler and not from the API.
When privacy is disabled we can return error code from the LE Set Random
Address HCI command instead.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Handle scanner active at RPA timeout

Fix RPA timeout handling when the scanner is active. An active scanner
must be restarted at RPA timeout otherwise the Set Random Address
command will fail.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -1180,6 +1232,10 @@ static void enh_conn_complete(struct bt_hci_evt_le_enh_conn_complete *evt)
 		conn = find_pending_connect(NULL);
 		if (!conn) {
+			if (IS_ENABLED(CONFIG_BT_WHITELIST)) {
+				notify_connected(NULL);
```
@jhedberg I'm a bit unsure about this one.

Maybe it would be better to allocate a valid conn object when starting whitelist conn create procedure, so that when we get connected we won't have to disconnect because we don't have enough conn objects.

jhedberg replied:
@joerchan we should pre-allocate a bt_conn any time that we start an action that may result in a connection. This includes any time we start connectable advertising, whitelist conn create procedure, directed advertising, etc. I had this on my mind already a long time ago, but I'm not sure I ever got around to opening an issue for it.

joerchan replied:
@jhedberg Alright. So we need connectable advertiser, and whitelist create connection. Directed advertising has it already. Possibly the old bt_le_set_auto_conn API.

Unrelated: I'm wondering if we maybe should deprecate bt_le_set_auto_conn  in favour of using whitelist.

joerchan commented:
Will continue with this after: https://github.com/zephyrproject-rtos/zephyr/pull/21743

joerchan committed:
Bluetooth: host: Set conn state before sending HCI command

Handle a possible race condition in the host connection state.
Set the conn state of the connection object before command is sent. This
is in case the calling function is not scheduled again before the
connection complete event arrives. In this case find_pending_conn will
not find the connection object.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Refactor initiator to common code and initiator state

Refactor the handling of sending the LE Create Connection command and
for whitelist and direct initiator to use the same host state flag
and common handling of the privacy address.
Also simplify the way we check if the procedure has already been started
when application tries to start it again.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Disconnect connection if no conn object is available

When receiving a connection complete event but no connection object are
available in the host something strange has happened. In this case
the controller might have a connection that cannot be controlled by the
application. It would then be sensible to disconnect this connection in
the controller.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Ifdef starting of background scanner in init

Starting a background scanner for auto-connection is an API that is only
available when whitelist API is not enabled.
There is currently no way to set this bit when the whitelist API is
enabled so there is not any issues with the current code, but it is
still not correct.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Handle initiator at RPA timeout

Handle initiator role when RPA timeout expires. For direct connect
establishment procedure we make sure the RPA is refreshed when starting
initiator and limit the timeout to the RPA timeout.
For auto establishment procedure we cancel the initiator and restart it
again in the connection complete event that is generated when canceling
an initiator.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: shell: Add shell command for advertise with identity option

Add shell command for advertise with identity option to have the
possibility to temporarily disable privacy feature in the advertise from
the shell.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Fix advertiser with identity switches to RPA

Fix advertiser requested to use the identity address while privacy
feature is enabled will change to using RPA address when advertise is
resumed or when RPA timeout occurred.
RPA timeout does not need to run when advertiser is using identity.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan commented:
@jhedberg @carlescufi This one is a bit difficult. And even now I have ignored the use of NRPA addresses.

NRPA is used when the privacy feature is disabled (BT_PRIVACY=n), in that case a non-connectable advertise or an active scanner will use an NRPA as it's address. I'm now very eager to simply remove this behavior from the stack as this address handling is already complicated enough. And I don't see why NRPA addresses should be used when privacy feature wasn't enabled to begin with.

jhedberg commented:
@joerchan short answer: mesh. Every packet goes out with unique NRPA and SMP support is disabled.

joerchan commented:
@jhedberg Thanks for the help in understand the MESH use-case. I realized that was overdoing the checking of incompatible use of the random address. So now I only check for when BT_PRIVACY is enabled. Since this is the only case where the existing code had an issue.

The issue was that using different ID for random address, could change the initiator or scanner RPA to the identity for the advertiser. 

Otherwise it should be the same now. 

This work has gone different places, and might not all be considered part of the same bug fix. The commits are hard to re-order right now, but if needed we can split the PR.

joerchan commented:
Had some trouble to avoid:
```
hci_core.c:556:13: error: 'le_update_private_addr' defined but not used [-Werror=unused-function]
 static void le_update_private_addr(void)
```
In the end I had to use `ifdef` even under `if (IS_ENABLED())`

jhedberg commented:
@joerchan are you sure? "not used" means that the code that's calling the function is behind #ifdef, so the alternative solution could be to remove #ifdef for that code

joerchan commented:
> @joerchan are you sure? "not used" means that the code that's calling the function is behind #ifdef, so the alternative solution could be to remove #ifdef for that code

@jhedberg Yes. Removing ifdef around rpa_timeout function I get the warning on `rpa_timeout` instead.
```
#if defined(CONFIG_BT_PRIVACY)
	k_delayed_work_init(&bt_dev.rpa_update, rpa_timeout);
#endif
```
This I cannot put under `IS_ENABLED()` because `bt_dev.rpa_update` is undef `#ifdef` again.

joerchan commented:
@jhedberg To clarify: First I got Undefined reference for central applications. And then I got defined but not used.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -536,41 +537,85 @@ static int le_set_private_addr(u8_t id)
+		set_le_scan_enable(BT_HCI_LE_SCAN_DISABLE);
+		scan_enabled = true;
+	}
+#endif
```
Could this be put behind `IS_ENABLED()`? You'd then need to move `set_le_scan_enable()` away from #ifdef (which in this case can be done since it doesn't look like it depends on conditional struct members or anything like that)

joerchan replied:
It could. Will include this as a refactor commit at the end. Since this pattern is present in code not part of this PR. (If needed I can include refactor at the beginning and amend this commit).

joerchan replied:
Unfortunately this requires more changes since all functions that call `le_set_scan_enabl()` is under some form of ifdef, and fixing them causes new dependencies.
Either SMP, PRIVACY, CONN, CENTRAL, OBSERVER will cause problems. Which means non-connectable advertiser will cause an `defined but not reference` warning here.

Removing this change again. Must be done as part of a larger refactor.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -536,41 +537,85 @@ static int le_set_private_addr(u8_t id)
+	if (scan_enabled) {
+		set_le_scan_enable(BT_HCI_LE_SCAN_ENABLE);
+	}
+#endif
```
Same here

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -1225,19 +1318,34 @@ static void enh_conn_complete(struct bt_hci_evt_le_enh_conn_complete *evt)
+					bt_conn_set_state(conn,
+							  BT_CONN_DISCONNECTED);
+				}
 #endif /* !defined(CONFIG_BT_WHITELIST) */
```
Can this be converted to use `IS_ENABLED()` I don't see any obvious dependency problems.

joerchan replied:
It has an indentation problem, and IS_ENABLED introduces a new indentation.
I can combine some of the outer if statements and use IS_ENABLED.

jhedberg replied:
The deep indentation is a problem already in itself - it's an indication that this function needs some redesign - possibly by refactoring into some smaller helper functions or rethinking the function flow. Either way, I'm fine leaving this to a later PR.

joerchan replied:
Pushed a refactor commit to this. If you don't like it I can remove it again.

joerchan committed:
Bluetooth: host: Refactor use of #ifdef to IS_ENABLED()

Refactor use of #ifdef to IS_ENABLED() pattern for handling cancellation
of outgoing connection attempt. Reduce the amount of indentation by
combining outer if-statements.
Move handling of canceled create connection into a helper function.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>
